core/src/serve_repair.rs

```diff
@@ -455,20 +470,72 @@ impl ServeRepair {
         stats.dropped_requests_load_shed += dropped_requests;
         stats.total_requests += total_requests;
 
-        let root_bank = self.bank_forks.read().unwrap().root_bank();
-        let epoch_staked_nodes = root_bank.epoch_staked_nodes(root_bank.epoch());
-        for reqs in reqs_v {
-            self.handle_packets(
-                ping_cache,
-                recycler,
-                blockstore,
-                reqs,
-                response_sender,
-                stats,
-                data_budget,
-                &epoch_staked_nodes,
-            );
+        let decode_start = Instant::now();
+        let mut decoded_reqs = Vec::default();
+        for batch in reqs_v {
+            for packet in &batch {
+                let request: RepairProtocol = match packet.deserialize_slice(..) {
+                    Ok(request) => request,
+                    Err(_) => {
+                        stats.err_malformed += 1;
+                        continue;
+                    }
+                };
+
+                let from_addr = packet.meta.socket_addr();
+                if !ContactInfo::is_valid_address(&from_addr, &socket_addr_space) {
+                    stats.err_malformed += 1;
+                    continue;
+                }
+
+                if request.supports_signature() {
+                    // collect stats for signature verification
+                    Self::verify_signed_packet(&my_id, packet, &request, stats);
+                } else {
+                    stats.unsigned_requests += 1;
+                }
+
+                if request.sender() == &my_id {
+                    stats.self_repair += 1;
+                    continue;
+                }
+
+                let stake = match epoch_staked_nodes
+                    .as_ref()
+                    .map(|nodes| nodes.get(request.sender()))
+                    .unwrap_or_default()
+                {
+                    Some(stake) => {
```

Comment:
I don't know if `epoch_staked_nodes` includes any zeros, probably not.
Nevertheless I guess it wouldn't hurt to be explicit here and don't rely on that presumption.
Something like:
```rust
let stake = epoch_staked_nodes
  .as_ref()
  .and_then(|stakes| staked.get(request.sender()))
  .unwrap_or_default();
if stake == 0 {
   // ...
} else {
   //...
}
```

Reply:
done